Specialize StepBy<Range(Inclusive)>

Emerentius opened the pull request:
Part of #51557, related to #43064, #31155

As discussed in the above issues, `step_by` optimizes very badly on ranges which is related to
1. the special casing of the first `StepBy::next()` call
2. the need to do 2 additions of `n - 1` and `1` inside the range's `next()`

This PR eliminates both by overriding `next()` to always produce the current element and also step ahead by `n` elements in one go. The generated code is much better, even identical in the case of a `Range` with constant `start` and `end` where `start+step` can't overflow. Without constant bounds it's a bit longer than the manual loop. `RangeInclusive` doesn't optimize as nicely but is still much better than the original asm.
Unsigned integers optimize better than signed ones for some reason.

See the following two links for a comparison.

[godbolt: specialization for ..](https://godbolt.org/g/haHLJr)
[godbolt: specialization for ..=](https://godbolt.org/g/ewyMu6)

`RangeFrom`, the only other range with an `Iterator` implementation can't be specialized like this without changing behaviour due to overflow. There is no way to save "finished-ness".

The approach can not be used in general, because it would produce side effects of the underlying iterator too early. 

May obsolete #51435, haven't checked.

Emerentius commented:
#36262 strikes again. And here I thought I'd be safe 'cause it's always clear which range type we have and all integers implement `Step`.

leonardo-m commented:
Making Specialize StepBy<Range(Inclusive)> efficient has a high priority in system language design.

rust-highfive commented:
The job `x86_64-gnu-llvm-3.9` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/393373857) ([raw log](https://api.travis-ci.org/v3/job/393373857/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[01:17:42] 
[01:17:42] running 845 tests
[01:17:42] ....................................................................................................
[01:17:42] ....................................................................................................
[01:17:42] ..........................................................F.........................................
[01:17:42] ....................................................................................................
[01:17:42] ....................................................................................................
[01:17:42] ....................................................................................................
[01:17:42] ....................................................................................................
[01:17:42] ....................................................................................................
[01:17:43] .............................................
[01:17:43] failures:
[01:17:43] 
[01:17:43] ---- iter::test_range_step stdout ----
[01:17:43] thread 'iter::test_range_step' panicked at 'assertion failed: `(left == right)`
[01:17:43]   left: `[200]`,
[01:17:43]  right: `[200, 250]`', libcore/../libcore/tests/iter.rs:1605:5
[01:17:43] 
[01:17:43] failures:
[01:17:43]     iter::test_range_step
[01:17:43] 
[01:17:43] 
[01:17:43] test result: FAILED. 842 passed; 1 failed; 2 ignored; 0 measured; 0 filtered out
[01:17:43] 
[01:17:43] error: test failed, to rerun pass '--test coretests'
[01:17:43] 
[01:17:43] 
[01:17:43] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-j" "4" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "-p" "core" "--" "--quiet"
[01:17:43] 
[01:17:43] 
[01:17:43] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:17:43] Build completed unsuccessfully in 0:29:51
[01:17:43] Build completed unsuccessfully in 0:29:51
[01:17:43] Makefile:58: recipe for target 'check' failed
[01:17:43] make: *** [check] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:2eb90abf
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Emerentius commented:
Pro-tip: if your local tests don't run against your own code, they are much more likely to succeed.

pietroalbini commented:
Picking someone from the libs team to review this. r? @sfackler

Emerentius commented:
I flipped the comparisons back around to keep the original behaviour for NaN-like values and squashed the commits together.
Updated and cleaned up the the godbolt code and exchanged the links in the OP. Signed integers don't optimize as well as unsigned ones, most likely because of the additional checks in `Step::add_usize` but it's still an improvement.

sfackler commented on src/libcore/iter/mod.rs:
```diff
@@ -737,6 +733,76 @@ impl<I> Iterator for StepBy<I> where I: Iterator {
 
+// hidden trait for specializing iterator methods
+// current use is limited to next() on StepBy
+trait SpecIter {
```
Could this trait mention `StepBy` in its name just to make it a bit more clear that it's limited to those impls?

sfackler commented on src/libcore/iter/mod.rs:
```diff
@@ -737,6 +733,76 @@ impl<I> Iterator for StepBy<I> where I: Iterator {
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.first_take = false;
+        if !(self.iter.start < self.iter.end) {
```
This comparison is structured weirdly to handle NaN properly?

Emerentius replied:
yes, it's `if start < end { return stuff } else { None }` flipped around so the positive case has no indent.

Emerentius commented:
@sfackler Renamed `SpecIter` to `StepBySpecIterator` and its `next()` method to `spec_next()` following `SpecExtend`.

sfackler commented:
@bors r+

bors commented:
:pushpin: Commit 000aff604e3b16ffc3771bd5d93a6e7b425852d2 has been approved by `sfackler`

<!-- @bors r=sfackler 000aff604e3b16ffc3771bd5d93a6e7b425852d2 -->

bors commented:
:hourglass: Testing commit 000aff604e3b16ffc3771bd5d93a6e7b425852d2 with merge 95979dcfb201f1d6906a2f256e5e311513cab4ea...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.7836), [status-travis](https://travis-ci.org/rust-lang/rust/builds/394920258?utm_source=github_status&utm_medium=notification)
Approved by: sfackler
Pushing 95979dcfb201f1d6906a2f256e5e311513cab4ea to master...

jethrogb commented:
This is buggy: https://github.com/rust-lang/rust/issues/55985

newpavlov commented:
So should we temporary revert this PR? Or should we leave it as-is and wait for the proper fix? I personally prefer the first option.

sfackler commented:
Let's revert until it's fixed, yeah.

leonardo-m commented:
The performance of this patch was not good enough :-(
